File: core/src/macbury/forge/screens/GameplayScreen.java
package macbury.forge.screens;

import com.badlogic.ashley.core.Entity;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input;
import com.badlogic.gdx.assets.loaders.ModelLoader;
import com.badlogic.gdx.graphics.g3d.Model;
import com.badlogic.gdx.graphics.g3d.ModelInstance;
import com.badlogic.gdx.graphics.g3d.loader.G3dModelLoader;
import com.badlogic.gdx.graphics.g3d.utils.FirstPersonCameraController;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.utils.UBJsonReader;
import macbury.forge.ForgE;
import macbury.forge.components.*;
import macbury.forge.db.models.Teleport;
import macbury.forge.level.Level;
import macbury.forge.utils.Vector3i;

/**
 * Created by macbury on 16.03.15.
 */
public class GameplayScreen extends AbstractScreen {
  private static final float FAR_CAMERA = 100;
  private static final float NEAR_CAMERA = 0.01f;
  private final Level level;
  private final Teleport teleport;
  private FirstPersonCameraController cameraController;

  public GameplayScreen(Teleport teleport, Level level) {
    this.level    = level;
    this.teleport = teleport;
  }

  @Override
  protected void initialize() {
    Gdx.input.setCursorCatched(true);
    this.cameraController     = new FirstPersonCameraController(level.camera);
    level.camera.far          = FAR_CAMERA;
    level.camera.near         = NEAR_CAMERA;
    level.camera.fieldOfView  = 70;
    Entity playerEntity       = ForgE.entities.get("player").build(level.entities);
    playerEntity.getComponent(PlayerComponent.class).camera = level.camera;
    level.terrainMap.localVoxelPositionToWorldPosition(teleport.voxelPosition, playerEntity.getComponent(PositionComponent.class).vector);
    playerEntity.getComponent(PositionComponent.class).vector.sub(-0.5f);
    level.entities.addEntity(playerEntity);

    level.entities.addEntity(ForgE.entities.get("teapot").build(level.entities));

  }

  @Override
  public void render(float delta) {
    ForgE.assets.loadPending();
    cameraController.update(delta);
    level.render(delta);

    if (Gdx.input.isKeyPressed(Input.Keys.ESCAPE)) {
      Gdx.app.exit();
    }
  }

  @Override
  public void resize(int width, int height) {
    level.resize(width, height);
  }

  @Override
  public void show() {

  }

  @Override
  public void hide() {

  }

  @Override
  public void pause() {

  }

  @Override
  public void resume() {

  }

  @Override
  public void dispose() {
    level.dispose();
  }
}

File: core/assets/db/scripts/screens/game_play_screen.rb
class GameplayScreen < AbstractScreen
  TAG                    = "GameplayScreen"
  FAR_CAMERA             = 70;
  NEAR_CAMERA            = 0.01;
  FIELD_OF_VIEW          = 70;
  attr_accessor :level, :teleport

  def onInitialize
    #Gdx.app.log(TAG, "Level=" + self.level.to_s)
    #Gdx.app.log(TAG, "Teleport=" + self.teleport.to_s)
    Gdx.input.setCursorCatched(true)
    @cameraController = FirstPersonCameraController.new(self.level.camera);

    self.level.camera.far          = FAR_CAMERA
    self.level.camera.near         = NEAR_CAMERA
    self.level.camera.fieldOfView  = FIELD_OF_VIEW

    @gameplay_view                 = GameplayView.new

    spawn_player
  end

  def spawn_player
    @playerEntity       = ForgE.entities.get("player").build(self.level.entities)
    @playerEntity.getComponent(PlayerComponent.java_class).camera = self.level.camera;
    self.level.terrainMap.localVoxelPositionToWorldPosition(self.teleport.voxelPosition, @playerEntity.getComponent(PositionComponent.java_class).vector);
    @playerEntity.getComponent(PositionComponent.java_class).vector.sub(-0.5);
    self.level.entities.addEntity(@playerEntity)
  end

  def spawn_entity
    boxEntity         = ForgE.entities.get("crate").build(self.level.entities)
    positionComponent = boxEntity.getComponent(PositionComponent.java_class)
    temp              = Vector3.new(@level.camera.direction).scl(2)
    temp.add(self.level.camera.position)
    positionComponent.vector.set(temp)
    temp.set(self.level.camera.direction).scl(15)
    boxEntity.getComponent(RigidBodyComoponent.java_class).initialImpulse.set(temp)
    self.level.entities.addEntity(boxEntity)
  end

  def render(delta)
    ForgE.time.update();
    ForgE.assets.update();
    @cameraController.update(delta)
    self.level.render(delta)

    spawn_entity if Gdx.input.isKeyJustPressed(Input::Keys::E)
    Gdx.app.exit if Gdx.input.isKeyPressed(Input::Keys::ESCAPE)

  end

  def resize(width, height)

  end

  def show
    ForgE.ui.addActor(@gameplay_view)
  end

  def hide
    @gameplay_view.remove
  end

  def pause

  end

  def resume

  end

  def dispose
    @playerEntity = nil
    self.level.dispose
    self.teleport = nil
    self.level    = nil
  end
end
